key_value_store.h

```diff
@@ -0,0 +1,30 @@
+#pragma once
+#include <iostream>
+#include <unordered_map>
+#include <mutex>
+#include <boost/functional/hash.hpp>
+
+namespace std {
+template <typename T, typename A> struct hash<std::vector<T, A> > {
+  size_t operator()(const std::vector<T, A>& v) const {
+    return boost::hash_range(v.cbegin(), v.cend());
+  }
+};
+}
+
+struct FlaggedValue {
+  std::vector<char> value;
+  uint32_t flags;
+};
+
+class KeyValueStore {
+  std::mutex mut;
+  std::unordered_map<std::vector<char>, FlaggedValue,
+                     std::hash<std::vector<char> > > data;
+
+public:
+  KeyValueStore() : mut() {}
```

Comment:
Comments! It's a public interface! What are these functions and how should I use them? Are there any edge cases I need to worry about?

Also, sometimes it's nice to use generous amounts of white space, or line up similar functions to make it easier to see differences. Again we're trying to reduce line noise so we don't just glaze over when looking at the code.
```c++
const FlaggedValue& get(const std::vector<char>& key);
void set(const std::vector<char>& key, const FlaggedValue& fv);
bool has(const std::vector<char>& key);
```